packages/samples/headless-react/src/components/search-box/search-box.function.tsx

```diff
@@ -0,0 +1,29 @@
+import { buildSearchBox, SearchBoxOptions } from "@coveo/headless";
+import { useEffect, useState } from "react";
+import { engine } from "../../engine";
+
+const options: SearchBoxOptions = {numberOfSuggestions: 8};
+const searchBox = buildSearchBox(engine, {options})
+
+export function SearchBox() {
+  const [state, setState] =  useState(searchBox.state);
+  const isEnterKey = (e: React.KeyboardEvent<HTMLInputElement>) => e.key === 'Enter';
+  
+  useEffect(() => searchBox.subscribe(() => setState(searchBox.state)), [])
+  
+  return (
+    <div>
+      <input
+      value={state.value}
+      onChange={(e) => searchBox.updateText(e.target.value)}
+      onKeyDown={(e) => isEnterKey(e) && searchBox.submit()}
+      />
+      <ul>
+        {state.suggestions.map((suggestion, index) => {
+          const value = suggestion.rawValue;
+          return <li key={index} onClick={() => searchBox.selectSuggestion(value)}>{value}</li>
+        })}
+      </ul>
+    </div>
+  );
+}
```

Comment:
I ran into infinite loops when instantiating a headless search-box controller inside React function components. There were two ways I could by-pass this, and both involved moving the controller outside the component.

1. Initialize the controller outside of the React function (show above for simplicity). This means that all `<SearchBox/>` instances are using the same headless search-box instance.
2. Initialize the controller outside of the React function, and pass the instance via a prop. This ensures every React instance has a unique search-box instance. Showing a sample of this though is more complex, and involves at least two files.

I do not think we should use option 1 since it would be misleading to devs. I wonder if there are better options to option 2.

Reply:
The second is pretty similar to what I did for the Pokemon challenge with headless. I put the engine inside its own context to make it available to all the sub components and then passed in the controllers as props.